tests/cpp/test_multidevice_preseg_passes.cpp

```diff
@@ -0,0 +1,186 @@
+// clang-format off
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2023-present NVIDIA CORPORATION & AFFILIATES.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+// clang-format on
+#include <gmock/gmock-matchers.h>
+#include <gtest/gtest.h>
+
+#include <fusion.h>
+#include <ops/all_ops.h>
+#include <preseg_passes/mark_aliases_prepare.h>
+#include <preseg_passes/optimization_pass.h>
+#include <preseg_passes/propagate_shardings.h>
+#include <runtime/fusion_executor_cache.h>
+#include <tests/cpp/multidevice.h>
+#include <tests/cpp/validator.h>
+#include "multidevice/utils.h"
+
+namespace nvfuser {
+
+constexpr int64_t b = 2, s = 3, h = 16, a = 2;
+constexpr double dropout_p = 0.0;
+constexpr bool is_causal = false;
+
+using testing::ElementsAre;
+using MultiDevicePresegPassesTest = MultiDeviceTest;
+
+TEST_F(MultiDevicePresegPassesTest, ResidualAdd) {
+  // This is similar to the residual add after MHA dropout in the transformer.
+  // The output of linear following MHA is all-gathered and sharded on the
+  // sequence dim. This sharding can be propagated to the linear output through
+  // backpropagating the shardings from residual add. This information is not
+  // present during forward propagation.
+  auto fusion = std::make_unique<Fusion>();
+  FusionGuard fg(fusion.get());
+
+  const int d = communicator_->size();
+
+  TensorView* tv0 = makeContigConcreteTensor({d, 4});
+  TensorView* tv1 = uniform(
+      shape(tv0),
+      fusion->zeroVal(DataType::Float),
+      fusion->oneVal(DataType::Float),
+      DataType::Float);
+  TensorView* tv2 = add(tv0, tv1);
+
+  auto mesh = DeviceMesh::createForNumDevices(d);
+  tv0->setDeviceMesh(mesh);
+  tv0->split(0, d, /*inner_split=*/false);
+  tv0->axis(0)->parallelize(ParallelType::DIDx);
+
+  fusion->addInput(tv0);
+  fusion->addOutput(tv1);
+  fusion->addOutput(tv2);
+
+  preseg_passes::OptimizationPass<
+      preseg_passes::PropagateShardingsPass>::runPass(fusion.get());
+  NVF_CHECK(tv1->hasDeviceMesh());
+  NVF_CHECK(
+      getShardedLogicalAxis(tv1, ParallelType::DIDx) ==
+          getShardedLogicalAxis(tv0, ParallelType::DIDx),
+      "Expected tv1 to be sharded like tv0 due to backpropagation of shardings.");
+}
+
+namespace {
+at::Tensor reference_mlp(at::Tensor inp, at::Tensor w0, at::Tensor w1) {
+  auto linear0 = at::linear(inp, w0);
+  auto gelu = at::gelu(linear0, "tanh");
+  auto linear1 = at::linear(gelu, w1);
+  return linear1;
+}
+
+at::Tensor reference_mha(at::Tensor inp) {
+  auto qkv = inp.transpose(1, 2).split(h / a, -1);
+  double scale = 1.0 / std::sqrt(h / a);
+  auto sdpa_out = at::_scaled_dot_product_flash_attention(
+      qkv[0],
+      qkv[1],
+      qkv[2],
+      /*dropout_p=*/dropout_p,
+      /*is_causal=*/is_causal,
+      /*return_debug_mask=*/false,
+      scale);
+  auto attn = std::get<0>(sdpa_out);
+  return attn.transpose(1, 2);
+}
+} // namespace
+
+TEST_F(MultiDevicePresegPassesTest, MLP) {
```

Comment:
AFAICT, this test does nothing in particular with preseg passes. They are run normally as part of runFusionWithInputs. If my above understanding is correct, I'd move it to test_multidevice_matmul.cpp or _transformer.cpp.